src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1431 @@
+
+    // verify the all_blocks list
+    debug_check_verify_all_blocks(provider, head);
+
+    // find the head (head->prev == NULL) of the upstream_alloc list
+    head = ravl_tree_get_head_block(provider->upstream_alloc);
+
+    // verify the upstream_alloc list
+    debug_check_verify_upstream_alloc(provider, head, &stats);
+
+    return true;
+}
+#endif
+
+static umf_result_t coarse_memory_provider_initialize(void *params,
+                                                      void **provider) {
+    umf_result_t umf_result;
+
+    if (provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (params == NULL) {
+        LOG_ERR("coarse provider parameters are missing");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    coarse_memory_provider_params_t *coarse_params =
+        (coarse_memory_provider_params_t *)params;
+
+    // check params
+    // we should either provider an upstream provider or init buffer
+    if (coarse_params->upstream_memory_provider == NULL &&
+        coarse_params->init_buffer == NULL) {
+        LOG_ERR("upstream provider or init buffer has to be provided in the "
+                "parameters");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (coarse_memory_provider_t *)umf_ba_global_alloc(
+            sizeof(coarse_memory_provider_t));
+    if (!coarse_provider) {
+        LOG_ERR("out of the host memory");
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    if (utils_mutex_init(&coarse_provider->lock) == NULL) {
+        LOG_ERR("lock initialization failed");
+        goto err_free_coarse_provider;
+    }
+
+    coarse_provider->upstream_memory_provider =
+        coarse_params->upstream_memory_provider;
+    coarse_provider->init_buffer = coarse_params->init_buffer;
+
+    coarse_provider->upstream_alloc =
+        ravl_new_sized(ravl_comp, sizeof(ravl_data_t));
+    if (coarse_provider->upstream_alloc == NULL) {
+        umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_free_coarse_provider;
+    }
+
+    coarse_provider->free_blocks =
+        ravl_new_sized(ravl_comp, sizeof(ravl_data_t));
+    if (coarse_provider->free_blocks == NULL) {
+        umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_delete_ravl_upstream_alloc;
+    }
+
+    coarse_provider->all_blocks =
+        ravl_new_sized(ravl_comp, sizeof(ravl_data_t));
+    if (coarse_provider->all_blocks == NULL) {
+        umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_delete_ravl_free_blocks;
+    }
+
+    coarse_provider->alloc_size = 0;
+    coarse_provider->used_size = 0;
+
+    if (coarse_params->immediate_init) {
+        // allocate and immediately deallocate memory using the upstream provider
+        void *init_buffer = NULL;
+        coarse_memory_provider_alloc(
+            coarse_provider, coarse_params->init_buffer_size, 0, &init_buffer);
```

Comment:
I think that a validation for `init_buffer_size = 0` could be useful (return `INVALID_ARGUMENT`).

Reply:
Done